ctapipe/image/extractor.py

```diff
@@ -1285,3 +1289,406 @@ def __call__(
+                    d_pk_idx = d.argmax()
+                    phase_pz2d_shifts.append(i_max - d_pk_idx)
+                    time = adaptive_centroid(
+                        d, d_pk_idx, leading_edge_rel_descend_limit
+                    )
+                else:
+                    time = i_max
+
+                phase_shifts.append(
+                    (time / upsampling * avg_step - ref_pulse_shape.argmax())
+                    * ref_sample_width_nsec
+                )
+                phase_gains.append(y[start:stop].sum() / integral)
+
+        if len(phase_gains) == 0:
+            raise ValueError(
+                "ref_pulse_shape is malformed - peak is not well contained"
+            )
+        gains.append(np.mean(phase_gains))
+        shifts.append(np.mean(phase_shifts))
+        pz2d_shifts.append(np.mean(phase_pz2d_shifts))
+
+    return pole_zeros, gains, shifts, pz2d_shifts
+
+
+def __filtfilt_fast(signal, filt):
+    """
+    Apply a linear filter forward and backward to a signal, based on scipy.signal.filtfilt.
+    filtfilt has some speed issues (https://github.com/scipy/scipy/issues/17080)
+    """
+    forward = convolve1d(signal, filt, axis=-1, mode="nearest")
+    backward = convolve1d(forward[::-1], filt, axis=-1, mode="nearest")
+    return backward[::-1]
+
+
+def deconvolve(
+    waveforms: npt.ArrayLike,
+    baselines: npt.ArrayLike,
+    upsampling: int,
+    pole_zero: float,
+) -> np.ndarray:
+    """
+    Applies pole-zero deconvolution and upsampling to pixel waveforms. Use
+    `deconvolution_parameters(...)` to estimate the required `pole_zero` parameter
+    for the specific camera model.
+
+    Parameters
+    ----------
+    waveforms : ndarray
+        Waveforms stored in a numpy array.
+        Shape: (n_pix, n_samples)
+    baselines : ndarray or float
+        Baseline estimates for each pixel that are subtracted from the waveforms
+        before deconvolution.
+        Shape: (n_pix, ) or scalar
+    upsampling : int
+        Upsampling factor to use (>= 1); if > 1, the input waveforms are resampled
+        at upsampling times their original sampling rate.
+    pole_zero : float
+        Deconvolution parameter obtained from `deconvolution_parameters(...)`.
+
+    Returns
+    -------
+    deconvolved_waveforms : ndarray
+        Deconvolved and upsampled waveforms stored in a numpy array.
+        Shape: (n_pix, upsampling * n_samples)
+    """
+    deconvolved_waveforms = np.atleast_2d(waveforms) - np.atleast_2d(baselines).T
+    deconvolved_waveforms[:, 1:] -= pole_zero * deconvolved_waveforms[:, :-1]
+    deconvolved_waveforms[:, 0] = 0
+
+    if upsampling > 1:
+        filt = np.ones(upsampling)
+        filt_weighted = filt / upsampling
+        signal = np.repeat(deconvolved_waveforms, upsampling, axis=-1)
+        return __filtfilt_fast(signal, filt_weighted)
+
+    return deconvolved_waveforms
+
+
+@guvectorize(
+    [
+        (float64[:], int64, float64, float32[:]),
```

Comment:
The guvectorize docs state:

> If there are several signatures, they must be ordered from the more specific to the least specific. Otherwise, Numpy’s type-based dispatching may not work as expected. For example, the following is wrong:
> ```@vectorize(["float64(float64)", "float32(float32)"])
def f(x): ...
```

So I think we need to switch the order here